examples/wmt/models.py

```diff
@@ -0,0 +1,774 @@
+        inputs_kv=x,
+        qkv_features=qkv_dim,
+        attention_axis=(1,),
+        causal_mask=False,
+        segmentation=inputs_segmentation,
+        padding_mask=padding_mask,
+        kernel_init=nn.initializers.xavier_uniform(),
+        bias_init=nn.initializers.normal(stddev=1e-6),
+        bias=False,
+        broadcast_dropout=False,
+        dropout_rate=attention_dropout_rate,
+        deterministic=deterministic)
+    x = nn.dropout(x, rate=dropout_rate, deterministic=deterministic)
+    x = x + inputs
+
+    # MLP block.
+    y = nn.LayerNorm(x, dtype=dtype)
+    y = MlpBlock(
+        y,
+        mlp_dim=mlp_dim,
+        dtype=dtype,
+        dropout_rate=dropout_rate,
+        deterministic=deterministic)
+
+    return x + y
+
+
+class EncoderDecoder1DBlock(nn.Module):
+  """Transformer encoder-decoder layer."""
+
+  def apply(self,
+            targets,
+            encoded,
+            qkv_dim,
+            mlp_dim,
+            num_heads,
+            dtype=jnp.float32,
+            inputs_segmentation=None,
+            targets_segmentation=None,
+            causal_mask=True,
+            padding_mask=None,
+            key_padding_mask=None,
+            dropout_rate=0.1,
+            attention_dropout_rate=0.1,
+            deterministic=False,
+            cache=None):
+    """Applies EncoderDecoder1DBlock module.
+
+    Args:
+      targets: input data for decoder
+      encoded: input data from encoder
+      qkv_dim: dimension of the query/key/value
+      mlp_dim: dimension of the mlp on top of attention block
+      num_heads: number of heads
+      dtype: the dtype of the computation (default: float32)
+      inputs_segmentation: input segmentation info for packed examples.
+      targets_segmentation: target segmentation info for packed examples.
+      causal_mask: bool, mask future or not
+      padding_mask: bool, mask padding tokens
+      key_padding_mask: bool, mask padding tokens
+      dropout_rate: dropout rate
+      attention_dropout_rate: dropout rate for attention weights
+      deterministic: bool, deterministic or not (to apply dropout)
+      cache: flax attention cache for fast decoding.
+
+    Returns:
+      output after transformer block.
+    """
+
+    # Decoder block.
+    assert targets.ndim == 3
+    x = nn.LayerNorm(targets, dtype=dtype)
+    x = nn.SelfAttention(
+        x,
+        num_heads=num_heads,
+        dtype=dtype,
+        inputs_kv=x,
+        qkv_features=qkv_dim,
+        attention_axis=(1,),
+        causal_mask=True,
+        padding_mask=padding_mask,
+        segmentation=targets_segmentation,
+        kernel_init=nn.initializers.xavier_uniform(),
+        bias_init=nn.initializers.normal(stddev=1e-6),
+        bias=False,
+        broadcast_dropout=False,
+        dropout_rate=attention_dropout_rate,
+        deterministic=deterministic,
+        cache=cache)
+    x = nn.dropout(x, rate=dropout_rate, deterministic=deterministic)
+    x = x + targets
```

Comment:
Why do we sum x with the targets here?

Reply:
This is the residual connection.  'targets' perhaps confusingly is the name I give for the inputs that arise from the target language in translation - perhaps I should come up w. better names for inputs/targets. - src_inputs, tgt_inputs..... not sure....